File: apps/api/app/schemas/task_description.py
from __future__ import annotations

import re
from typing import Annotated, Any, Literal

from pydantic import BaseModel, Field

CHECKBOX_LINE_PATTERN = re.compile(r"^\s*[-*]\s+\[( |x|X)\]\s*(.*)$")


class DescriptionTextBlock(BaseModel):
    kind: Literal["text"]
    text: str = ""


class DescriptionCheckboxBlock(BaseModel):
    kind: Literal["checkbox"]
    text: str = ""
    checked: bool = False


TaskDescriptionBlock = Annotated[
    DescriptionTextBlock | DescriptionCheckboxBlock,
    Field(discriminator="kind"),
]


def parse_legacy_description(description: str | None) -> list[dict[str, Any]]:
    if not description:
        return []

    lines = description.replace("\r\n", "\n").replace("\r", "\n").split("\n")
    blocks: list[dict[str, Any]] = []

    for line in lines:
        match = CHECKBOX_LINE_PATTERN.match(line)
        if match:
            blocks.append(
                {
                    "kind": "checkbox",
                    "text": match.group(2),
                    "checked": match.group(1).lower() == "x",
                }
            )
            continue

        blocks.append({"kind": "text", "text": line})

    return blocks


def normalize_description_blocks(
    blocks: list[TaskDescriptionBlock] | list[dict[str, Any]] | None,
) -> list[dict[str, Any]]:
    if not blocks:
        return []

    normalized: list[dict[str, Any]] = []

    for block in blocks:
        data = block.model_dump() if isinstance(block, BaseModel) else dict(block)
        kind = "checkbox" if data.get("kind") == "checkbox" else "text"
        text = str(data.get("text", "")).replace("\r\n", "\n").replace("\r", "\n")
        lines = text.split("\n")

        for line in lines:
            if kind == "checkbox":
                normalized.append(
                    {
                        "kind": "checkbox",
                        "text": line,
                        "checked": bool(data.get("checked", False)),
                    }
                )
                continue

            normalized.append({"kind": "text", "text": line})

    return normalized


def serialize_description_blocks(blocks: list[dict[str, Any]] | None) -> str | None:
    if not blocks:
        return None

    lines: list[str] = []

    for block in blocks:
        if block.get("kind") == "checkbox":
            marker = "x" if bool(block.get("checked", False)) else " "
            lines.append(f"- [{marker}] {str(block.get('text', ''))}")
            continue

        lines.append(str(block.get("text", "")))

    serialized = "\n".join(lines)
    return serialized if serialized.strip() else None

File: apps/web/src/lib/task-description.ts
import type { DescriptionBlock } from "@/lib/types";

export function createEmptyDescription(): DescriptionBlock[] {
  return [{ kind: "text", text: "" }];
}

export function normalizeDescriptionBlocks(blocks: DescriptionBlock[]): DescriptionBlock[] {
  if (blocks.length === 0) {
    return createEmptyDescription();
  }

  return blocks.flatMap((block) => {
    const normalizedLines = block.text.replace(/\r\n/g, "\n").replace(/\r/g, "\n").split("\n");

    return normalizedLines.map((line) =>
      block.kind === "checkbox"
        ? { kind: "checkbox" as const, text: line, checked: block.checked }
        : { kind: "text" as const, text: line },
    );
  });
}

export function stripDescriptionBlocks(blocks: DescriptionBlock[]): DescriptionBlock[] {
  const normalized = normalizeDescriptionBlocks(blocks);
  const trimmedEndIndex = [...normalized]
    .reverse()
    .findIndex((block) => block.text.trim().length > 0 || (block.kind === "checkbox" && block.checked));

  if (trimmedEndIndex === -1) {
    return [];
  }

  return normalized.slice(0, normalized.length - trimmedEndIndex);
}

export function descriptionBlocksToText(blocks: DescriptionBlock[]): string | null {
  const normalized = stripDescriptionBlocks(blocks);
  if (normalized.length === 0) {
    return null;
  }

  const serialized = normalized
    .map((block) =>
      block.kind === "checkbox" ? `- [${block.checked ? "x" : " "}] ${block.text}` : block.text,
    )
    .join("\n");

  return serialized.trim().length > 0 ? serialized : null;
}

export function descriptionTextToBlocks(description: string | null): DescriptionBlock[] {
  if (!description) {
    return createEmptyDescription();
  }

  const blocks = description.split(/\r?\n/).map((line) => {
    const checkboxMatch = line.match(/^\s*[-*]\s+\[( |x|X)\]\s*(.*)$/);
    if (checkboxMatch) {
      return {
        kind: "checkbox" as const,
        text: checkboxMatch[2],
        checked: checkboxMatch[1].toLowerCase() === "x",
      };
    }

    return { kind: "text" as const, text: line };
  });

  return blocks.length > 0 ? blocks : createEmptyDescription();
}

export function ensureDescriptionBlocks(blocks: DescriptionBlock[] | null | undefined, fallback?: string | null) {
  if (blocks && blocks.length > 0) {
    return normalizeDescriptionBlocks(blocks);
  }

  return descriptionTextToBlocks(fallback ?? null);
}

export function hasMeaningfulDescription(blocks: DescriptionBlock[] | null | undefined, fallback?: string | null): boolean {
  const normalizedBlocks = ensureDescriptionBlocks(blocks, fallback);
  return stripDescriptionBlocks(normalizedBlocks).some(
    (block) => block.text.trim().length > 0 || (block.kind === "checkbox" && block.checked),
  );
}
